### DISCUSSION TITLE
How to access app handler from an invoke function?

### DISCUSSION THREAD

**Discussion by abisar** [tauri-apps/tauri#4729] [Q&A] [answered]
How do I change tray icon from this `r_stuff` function?
```
#[tauri::command(async)]
fn r_stuff() {
    // How to do I change tray icon or access the app handler from here?  <----------
    app.tray_handle().set_icon(tauri::Icon::Raw(include_bytes!("../path/to/myicon.ico"))).unwrap();
}


fn main() {
    let context = tauri::generate_context!();

    let quit = CustomMenuItem::new("quit".to_string(), "Quit");
    let show = CustomMenuItem::new("show".to_string(), "Show app");
    let tray_menu = SystemTrayMenu::new()
        .add_item(show)
        .add_native_item(SystemTrayMenuItem::Separator)
        .add_item(quit);
    let tray = SystemTray::new().with_menu(tray_menu);

    tauri::Builder::default()
        .system_tray(tray)
        .on_system_tray_event(|app, event| match event {...})
        .invoke_handler(tauri::generate_handler![r_stuff])
        .run(context)
        .expect("Error while running tauri application");
}
```

I don't understand [this part](https://tauri.app/v1/guides/features/system-tray/#updating-system-tray) of the documentation:
![image](https://user-images.githubusercontent.com/18389149/180367221-6bddc112-d58a-4c35-9650-19c7c8b811cb.png)

**Comment by vdvman1** [comment, marked as answer]
If you look at the official command guide there is a section specifically on how to get an `AppHandle` in a command [https://tauri.app/v1/guides/features/command#accessing-an-apphandle-in-commands](https://tauri.app/v1/guides/features/command#accessing-an-apphandle-in-commands)

**Reply by abisar** [reply]
Thank you! I totally missed it.

**Reply by juancarloselorriaga** [reply]
Thanks it was helpful for me too. One other question. This is the app handle, is there any way to access the app from a command?

**Reply by FabianLars** [reply]
No, the App can realistically only be accessed in the setup hook. _but_ the AppHandle is supposed to be able to do everything the App instance can do (it's a handle to it after all) - what are you missing from it?

**Reply by juancarloselorriaga** [reply]
I'm setting up a config file with [tauri-plugin-store](https://github.com/tauri-apps/tauri-plugin-store). 

I'm creating the store inside the setup hook with no problem.

```rs
 .setup(|app| {
            let stores = app.state::<StoreCollection<Wry>>();
            let path = PathBuf::from("settings.dat");
            with_store(app.handle(), stores, path, |store| {
                // some additional logic here
                Ok(())
            })?;

            Ok(())
        })

```

But I can't find an easy way to access this specific information from a command. If I create a new store with the same name this one will be empty, and the keys added in the setup hook aren't accessible. If I try to use the with_store hook from a command, I don't know if the same state is accessible from the command args, I've tried that too.

So most probably I'm missing something pretty obvious here, but I definitely need help to make this work. My end goal is basically access the same store from the setup hook (to set it up) and from a command (rust) and of course from JS for reading and updating purposes from the FE.

**Reply by FabianLars** [reply]
So i'm a bit rusty when it comes to the store plugin, but as far as i know the idea is to do basically the same thing as you did in the code you posted:
```rs
use tauri::{Manager, Wry};
use tauri_plugin_store::with_store;

#[tauri::command]
fn do_whatever(app: tauri::AppHandle) {
  let stores = app.state::<StoreCollection<Wry>>();
  let path = PathBuf::from("settings.dat");
  with_store(app.handle(), stores, path, |store| {
    // some additional logic here
    Ok(())
  })?;
}
```
and it's supposed to have the values you set in the setup hook.

**Reply by juancarloselorriaga** [reply]
Oops! Thank you very much, it works as expected!
